File: majorProject/src/main/java/com/aocc/framework/implementation/SingleTouchHandler.java
package com.aocc.framework.implementation;

import java.util.ArrayList;
import java.util.List;

import android.view.MotionEvent;
import android.view.View;

import com.aocc.framework.Pool;
import com.aocc.framework.Input.TouchEvent;
import com.aocc.framework.Pool.PoolObjectFactory;
import com.aocc.framework.Viewport;

public class SingleTouchHandler implements TouchHandler {
    boolean isTouched;
    int touchX;
    int touchY;
    Pool<TouchEvent> touchEventPool;
    List<TouchEvent> touchEvents = new ArrayList<TouchEvent>();
    List<TouchEvent> touchEventsBuffer = new ArrayList<TouchEvent>();
    Viewport viewport;

    public SingleTouchHandler(View view, Viewport viewport) {
        PoolObjectFactory<TouchEvent> factory = new PoolObjectFactory<TouchEvent>() {
            @Override
            public TouchEvent createObject() {
                return new TouchEvent();
            }
        };
        touchEventPool = new Pool<TouchEvent>(factory, 100);
        view.setOnTouchListener(this);
        this.viewport = viewport;
    }

    @Override
    public void setViewport(Viewport viewport) {
        this.viewport = viewport;
    }

    private int toWorldX(float screenX) {
        return viewport != null ? viewport.screenToWorldX(screenX) : (int) screenX;
    }

    private int toWorldY(float screenY) {
        return viewport != null ? viewport.screenToWorldY(screenY) : (int) screenY;
    }

    @Override
    public boolean onTouch(View v, MotionEvent event) {
        synchronized(this) {
            TouchEvent touchEvent = touchEventPool.newObject();
            switch (event.getAction()) {
            case MotionEvent.ACTION_DOWN:
                touchEvent.type = TouchEvent.TOUCH_DOWN;
                isTouched = true;
                break;
            case MotionEvent.ACTION_MOVE:
                touchEvent.type = TouchEvent.TOUCH_DRAGGED;
                isTouched = true;
                break;
            case MotionEvent.ACTION_CANCEL:
            case MotionEvent.ACTION_UP:
                touchEvent.type = TouchEvent.TOUCH_UP;
                isTouched = false;
                break;
            }

            touchEvent.x = touchX = toWorldX(event.getX());
            touchEvent.y = touchY = toWorldY(event.getY());
            touchEventsBuffer.add(touchEvent);

            return true;
        }
    }

    @Override
    public boolean isTouchDown(int pointer) {
        synchronized(this) {
            if(pointer == 0)
                return isTouched;
            else
                return false;
        }
    }

    @Override
    public int getTouchX(int pointer) {
        synchronized(this) {
            return touchX;
        }
    }

    @Override
    public int getTouchY(int pointer) {
        synchronized(this) {
            return touchY;
        }
    }

    @Override
    public List<TouchEvent> getTouchEvents() {
        synchronized(this) {
            int len = touchEvents.size();
            for( int i = 0; i < len; i++ )
                touchEventPool.free(touchEvents.get(i));
            touchEvents.clear();
            touchEvents.addAll(touchEventsBuffer);
            touchEventsBuffer.clear();
            return touchEvents;
        }
    }
}

File: majorProject/src/main/kotlin/com/aocc/framework/implementation/SingleTouchHandler.kt
package com.aocc.framework.implementation

import android.view.MotionEvent
import android.view.View
import com.aocc.framework.Input
import com.aocc.framework.Pool
import com.aocc.framework.Viewport

class SingleTouchHandler(view: View, viewport: Viewport) : TouchHandler {

    private var isTouched: Boolean = false
    private var touchX: Int = 0
    private var touchY: Int = 0
    private val touchEventPool: Pool<Input.TouchEvent>
    private val touchEventsList = ArrayList<Input.TouchEvent>()
    private val touchEventsBuffer = ArrayList<Input.TouchEvent>()
    private var viewport: Viewport? = viewport

    init {
        touchEventPool = Pool(Pool.PoolObjectFactory { Input.TouchEvent() }, 100)
        view.setOnTouchListener(this)
    }

    override fun setViewport(viewport: Viewport) {
        this.viewport = viewport
    }

    private fun toWorldX(screenX: Float): Int {
        return viewport?.screenToWorldX(screenX) ?: screenX.toInt()
    }

    private fun toWorldY(screenY: Float): Int {
        return viewport?.screenToWorldY(screenY) ?: screenY.toInt()
    }

    override fun onTouch(v: View, event: MotionEvent): Boolean {
        synchronized(this) {
            val touchEvent = touchEventPool.newObject()
            when (event.action) {
                MotionEvent.ACTION_DOWN -> {
                    touchEvent.type = Input.TouchEvent.TOUCH_DOWN
                    isTouched = true
                }
                MotionEvent.ACTION_MOVE -> {
                    touchEvent.type = Input.TouchEvent.TOUCH_DRAGGED
                    isTouched = true
                }
                MotionEvent.ACTION_CANCEL, MotionEvent.ACTION_UP -> {
                    touchEvent.type = Input.TouchEvent.TOUCH_UP
                    isTouched = false
                }
            }

            touchEvent.x = toWorldX(event.x).also { touchX = it }
            touchEvent.y = toWorldY(event.y).also { touchY = it }
            touchEventsBuffer.add(touchEvent)

            return true
        }
    }

    override fun isTouchDown(pointer: Int): Boolean {
        synchronized(this) {
            return pointer == 0 && isTouched
        }
    }

    override fun getTouchX(pointer: Int): Int {
        synchronized(this) {
            return touchX
        }
    }

    override fun getTouchY(pointer: Int): Int {
        synchronized(this) {
            return touchY
        }
    }

    override val touchEvents: List<Input.TouchEvent>
        get() {
            synchronized(this) {
                val len = touchEventsList.size
                for (i in 0 until len) {
                    touchEventPool.free(touchEventsList[i])
                }
                touchEventsList.clear()
                touchEventsList.addAll(touchEventsBuffer)
                touchEventsBuffer.clear()
                return touchEventsList
            }
        }
}
